tools/clang/unittests/HLSL/ExecutionTest.cpp

```diff
@@ -8929,15 +9504,15 @@ void ExecutionTest::RunResourceTest(ID3D12Device *pDevice, const char *pShader,
   // Create SRVs
   CreateRawSRV(pDevice, baseHandle, valueSize, pSRVResources[0]);
   CreateStructSRV(pDevice, baseHandle, valueSize, sizeof(float), pSRVResources[1]);
-  CreateTex2DSRV(pDevice, baseHandle, valueSize, DXGI_FORMAT_R32_FLOAT, pSRVResources[2]);
+  CreateTex2DSRV(pDevice, baseHandle, DXGI_FORMAT_R32_FLOAT, pSRVResources[2]);
   // Create UAVs
   CreateRawUAV(pDevice, baseHandle, valueSize, pUAVResources[0]);
   CreateStructUAV(pDevice, baseHandle, valueSize, sizeof(float), pUAVResources[1]);
   CreateTypedUAV(pDevice, baseHandle, valueSize, DXGI_FORMAT_R32_FLOAT, pUAVResources[2]);
   CreateTex1DUAV(pDevice, baseHandle, DXGI_FORMAT_R32_FLOAT, pUAVResources[3]);
 
   D3D12_FILTER filters[] = {D3D12_FILTER_MIN_MAG_LINEAR_MIP_POINT, D3D12_FILTER_COMPARISON_MIN_MAG_LINEAR_MIP_POINT};
-  float borderColors[] = {30.0, 31.0};
+  float borderColors[] = {30.0, 31.0, 0.0, 0.0};
```

Comment:
It's not obvious, but borderColors is indexed by sampler (count = NumSamplers), not RGBA index.  So, what's the purpose of expanding this array?  Though this *is* a bit confusing when using this "CreateDefaultSamplers" function.  Perhaps `borderColors` should be renamed to `perSamplerBorderColorScalars`, or some such.

Reply:
`D3D12_SAMPLER_DESC` has the element `BorderColor` which is an array of 4. `CreateDefaultSamplers` tried to initialize all four from this parameter, leading to garbage that didn't harm anything to my knowledge, but it was wrong.

I don't think I follow what you're saying about indexing by sampler and not RGBA index. I'm happy to rename it, but I don't think I see the confusion yet. Probably because I created the function 😓

Reply:
Because this array doesn't map to the 4 BorderColor array elements in D3D12_SAMPLER_DESC.  In `CreateDefaultSamplers`, this array is interpreted as an array of scalars, one per sampler.

In this assignment: `sampDesc.BorderColor[j] = borderColors[i];`, `i` is the sampler index, while `j` is the RGBA BorderColor array index.  Each scalar is replicated across all RGBA components for one sampler's BorderColor.  This code should not have been reading off the end of the array with `NumSamplers == 2`.

That's what's confusing about the name.  Intuitively, this looks like it would be RGBA, but it's not used that way.

Reply:
ah yes! I misunderstood my own code. I'll fix it